### ISSUE TITLE
Add option to depends_on to automatically propagate package variants to dependencies

### ISSUE THREAD

**Issue by kwryankrattiger** [LLNL/spack#25545]
### Summary

I want a simpler way to propagate variants between packages. The default behavior is to not do this, which makes sense, but in many cases it is desired to propagate variants to dependencies.

### Rationale

Currently propagating package variants is a bit of a pain as each flag must be individually specified on and off.

```python
depends_on('hdf5+mpi', when='+hdf5+mpi')
depends_on('hdf5~mpi', when='+hdf5~mpi')
```

This becomes very verbose as the number of variants increases. And with the old concretizer it is even more verbose as it is required to use the Cartesian product (ref. https://github.com/spack/spack/blob/c4e50c9efbc0c99ae272e27387a33fdc5f4aea23/var/spack/repos/builtin/packages/ecp-data-vis-sdk/package.py#L47).

### Description

A possible solution to this is currently being implemented to replace the method mentioned above. It is incomplete as it is now for general use.

https://github.com/kwryankrattiger/spack/blob/42453310b81ce8c1ce7c29bbd0ee031a086a1abe/var/spack/repos/builtin/packages/ecp-data-vis-sdk/package.py#L49

The idea here is to pass a list or map of variants to be propagated and the function takes care of adding the appropriate variations. It is far less verbose and is clear from the signature what is happening.

### Additional information

Basing this on the latest develop as of Aug 20.

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by haampie** [comment]
You can do

```
with when('+hdf5'):
    depends_on('hdf5+mpi', when='+mpi')
    depends_on('hdf5~mpi', when='~mpi')
```

do you really explicitly depend on hdf5~mpi though?

Generally it's better to keep the package.py as declarative as possible

**Comment by kwryankrattiger** [comment]
In the case that I am describing we are explicitly turning the variant on/off based on the variant from the parent. The use case is more wrapper packages that setup SDKs vs individual packages. ie. the SDK doesn't have MPI, all of the sub packages turn off MPI.

I agree too much magic can lead to issues. The proposed solution is much less verbose and it is a little easier to parse one line vs many, especially when the many lines look very similar to each other differing by only one character. So for that reason I think there is a benefit. But it is also pretty easy to just write a wrapper for the packages that need it so it isn't the end of the world.

```python
with when('+hdf5'):
  depends_on('hdf5+mpi', when='+mpi')
  depends_on('hdf5~mpi', when='~mpi')
  depends_on('hdf5+fortran', when='+fortran')
  depends_on('hdf5~fortran', when='~fortran')
```

```python
depends_on('hdf5', when='+hdf5', propagate=['mpi','fortran'])
```

**Comment by haampie** [comment]
Yeah, that looks kinda useful, it could also make sense for propagating say `cuda_arch`

**Comment by haampie** [comment]
What syntax would make sense if versions were propagated? That's also common e.g. in the rocm ecosystem, where currently loops are used
